pkg/stub/handler.go

```
@@ -0,0 +1,176 @@
+	"k8s.io/client-go/dynamic"
+	corev1 "k8s.io/api/core/v1"
+	"errors"
+)
+
+func NewHandler(m *Metrics, osClient openshift.OSClient, dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)) sdk.Handler {
+	return &Handler{
+		metrics: m,
+		osClient: osClient,
+		dynamicResourceClientFactory: dynamicResourceClientFactory,
+	}
+}
+
+type Metrics struct {
+	operatorErrors prometheus.Counter
+}
+
+type Handler struct {
+	// Metrics example
+	metrics *Metrics
+	osClient openshift.OSClient
+	dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)
+}
+
+func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+	switch o := event.Object.(type) {
+	case *v1alpha1.WebApp:
+		if o.GetDeletionTimestamp() != nil {
+			err := h.delete(o)
+			if err != nil {
+				logrus.Errorf("Error deleting all operator related resources: %v", err)
+				h.setStatus(err.Error(), o)
+				return err
+			}
+			return nil
+		}
+		exts, err := h.processTemplate(o)
+		if err != nil {
+			logrus.Errorf("Error while processing the template: %v", err)
+			h.setStatus(err.Error(), o)
+			return err
+		}
+
+		runtimeObjs, err := h.getRuntimeObjs(exts)
+		if err != nil {
+			logrus.Errorf("Error parsing the runtime objects from the template: %v", err)
+			h.setStatus(err.Error(), o)
+			return err
+		}
+
+		err = h.provisonObjects(runtimeObjs, o)
+		if err != nil {
+			logrus.Errorf("Error provisioning the runtime objects: %v", err)
+			h.setStatus(err.Error(), o)
+			return err
+		}
+
+		if h.isAppReady(o) {
+			h.setStatus("OK", o)
+		} else {
+			h.setStatus("", o)
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func (h *Handler) delete(cr *v1alpha1.WebApp) error {
+	ns := cr.Namespace
+	appName :=  cr.Spec.DCName
+
+	return h.osClient.Delete(ns, appName)
+}
+
+func (h *Handler) setStatus(msg string, cr *v1alpha1.WebApp) {
+	cr.Status.Message = msg
+	sdk.Update(cr)
```

ignoring an error here